### DISCUSSION TITLE
CliRunner for non-Click projects

### DISCUSSION THREAD

**Discussion by RhetTbull** [pallets/click#2640] [Ideas]
I love Click and usually reach for it when building a CLI. However, I'll occasionally use `argparse` for projects where I don't want to introduce another dependency. For example, projects where the CLI is ancillary to the main project (e.g. providing an interface via `python -m packagename`).  In these cases, I really miss the excellent `CliRunner` provided by Click for unit testing. To solve this, I've ported Click's CliRunner to a standalone project, [clirunner](https://github.com/RhetTbull/clirunner), and modified it to work with "regular" CLIs built with `argparse` or manual `sys.argv` parsing.

The interface is the same as Click's CliRunner minus any Click-specific things. Behavior is the same with two exceptions: 

1. `stderr` is flushed on exit so that you can test `print("Foo", file=sys.stderr)` type statements. Click doesn't do this which changes the expected behavior as `stderr` is normally line-buffered in the standard library.  This is fine for Click apps using `echo()` which does flush the buffer but for non-Click apps, this could result in unexpected behavior. 
2. The `stdout` and `stderr` properties of the `Result` object are always available and `mix_stderr` option is removed. This matches the desired behavior for #2522 for which there is an open pull-request.

[clirunner.CliRunner](https://github.com/RhetTbull/clirunner) should make it easier to test these "non-Click" projects while giving you an easy upgrade path if you do later to decide to use Click as the tests will be the same.

**Comment by davidism** [comment]
Cool! Be sure to include our license alongside your own in the source, sdist, and wheel.

**Reply by RhetTbull** [reply]
It is. The LICENSE file is included in all the distribution files as:

```
Copyright (c) 2023, Rhet Turnbull

Adapted from the Click project, copyright 2014 Pallets

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are
met:

1.  Redistributions of source code must retain the above copyright
    notice, this list of conditions and the following disclaimer.

2.  Redistributions in binary form must reproduce the above copyright
    notice, this list of conditions and the following disclaimer in the
    documentation and/or other materials provided with the distribution.

3.  Neither the name of the copyright holder nor the names of its
    contributors may be used to endorse or promote products derived from
    this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
```

**Reply by davidism** [reply]
It will need to be a separate file, from what I understand.

**Reply by RhetTbull** [reply]
OK, I'll add a separate file but I don't read the license that way. It states "Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer..." which is all included in the projects LICENSE including the "copyright 2014 Pallets" notice so I think the intent is met.  But no worries, I'll add a separate file.

**Reply by RhetTbull** [reply]
Done!

**Comment by davidism** [comment]
Yeah, I'm not an expert at interpreting these, but thanks!
